protocols/relay/src/v2/client.rs

```diff
@@ -0,0 +1,412 @@
+        _handler: handler::Handler,
+    ) {
+        self.connected_peers.get_mut(peer_id).map(|cs| {
+            cs.remove(
+                cs.iter()
+                    .position(|c| c == connection_id)
+                    .expect("Connection to be known."),
+            )
+        });
+    }
+
+    fn inject_event(
+        &mut self,
+        event_source: PeerId,
+        _connection: ConnectionId,
+        handler_event: handler::Event,
+    ) {
+        match handler_event {
+            handler::Event::ReservationReqAccepted { renewal } => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::ReservationReqAccepted {
+                            relay_peer_id: event_source,
+                            renewal,
+                        },
+                    ))
+            }
+            handler::Event::ReservationReqFailed { renewal } => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::ReservationReqFailed {
+                            relay_peer_id: event_source,
+                            renewal,
+                        },
+                    ))
+            }
+            handler::Event::OutboundCircuitReqFailed {} => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::OutboundCircuitReqFailed {
+                            relay_peer_id: event_source,
+                        },
+                    ))
+            }
+            handler::Event::InboundCircuitReqDenied { src_peer_id } => self
+                .queued_actions
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    Event::InboundCircuitReqDenied { src_peer_id },
+                )),
+            handler::Event::InboundCircuitReqDenyFailed { src_peer_id, error } => self
+                .queued_actions
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    Event::InboundCircuitReqDenyFailed { src_peer_id, error },
+                )),
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        _poll_parameters: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        if let Some(event) = self.queued_actions.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        loop {
+            match self.from_transport.poll_next_unpin(cx) {
+                Poll::Ready(Some(transport::TransportToBehaviourMsg::ListenReq {
+                    relay_peer_id,
+                    relay_addr,
+                    to_listener,
+                })) => {
+                    match self
+                        .connected_peers
+                        .get(&relay_peer_id)
+                        .and_then(|cs| cs.get(0))
```

Comment:
Today rust-libp2p has no active mechanism to close duplicate connections to the same peer. E.g. say that one establishes two TCP connections to the same peer, no logic within `libp2p-swarm` or `libp2p-core` will remove one of the two for the sake of reducing used resources.

Closing a connection in coordination with a remote is actually a bit complicated. See https://github.com/libp2p/specs/issues/328 for a proposal of a graceful migration to a single connection.

What rust-libp2p does offer today is a mechanism to close idle connections, that mechanism being `ProtocolsHandler::keep_alive`.

Say that the local node has multiple connections to the same peer, with one of the two being created distinguishably after the other. If both peers always use the first connection, the second will go idle and will eventually be closed.

Now this is not ideal as (a) it is not specified anywhere (b) it depends on both peers seeing the same connection as _the earlier one_ and (c) doesn't take any properties of the connection into consideration (e.g. bandwidth or a direct over a relayed connection).

Does the above make sense @elenaf9? Do you agree that, even though not ideal, it is still a mechanism worth using?

Reply:
Okay yes, thanks for the clarification. Since we then always use the same connection on each reservation request, couldn't we check in `client::Handler::inject_event(In::Reserve)` if we already have a reservation, and only send a `outbound_hop::Reserve` if we don't have one yet?

Reply:
I would deem the case where one calls `listen_on` with the same relay address more than once unlikely.

As of today, correct me if I am wrong, in case one would call `listen_on` with the same relay address twice, the following would happen:

- The local node requests a new reservation with the relay.
- The relay server should see this as a reservation refresh.
- Once accepted, the local node replaces the existing reservation with the new reservation.
  - Thus dropping the sender to the old listener which in turn closes.
  - Sending a `transport::Reservation` to the new listener.

We could instead not request a new reservation, still drop the old listener and notify the new listener. Under the above assumption that it is unlikely for this to occur in the first place and given that the current logic still behaves correctly (ignoring the additional roundtrip to the relay for a new reservation to the relay), I would leave it as is for the sake of less lines of code.

Would you agree with the reasoning above?

Reply:
> Thus dropping the sender to the old listener which in turn closes.

Doesn't this actually causes the listener to panic?
https://github.com/mxinden/rust-libp2p/blob/7b2eeb5468ba8b875056fdafada6950a52d8277f/protocols/relay/src/v2/client/transport.rs#L333

Reply:
True. Sorry, should have caught this before. a0750efd6d515da9b0eefd3fa99f7d5b0885f980 adds a test triggering the panic and 4b90626ab232faacba28139471f7c6fba0b520ab fixes the panic closing the listener. The listener closes without an error (`Poll::Ready(None)`) as I would expect the old listener closing in favor of the new to be intended by the user. Happy for alternative suggestions.

With these two commits, do you think this use case (double listening via same relay) is covered?

---

Currently the listener would as well close when the connection to the relay closes. I am not sure whether retries should be handled within `libp2p-relay` or outside by calling `Swarm::listen_on` again. Still need to give this more thought. Adding an item to my TODO list. Likely worth delaying till we implement libp2p-autorelay` combining both `libp2p-relay` and `libp2p-autonat`.

Reply:
> With these two commits, do you think this use case (double listening via same relay) is covered?

Yes, I agree that it is only a rare case, and the current implementation is fine.

> Likely worth delaying till we implement libp2p-autorelay combining both libp2p-relayandlibp2p-autonat`.

Doing this via a autorelay protocol sounds good.